### PR TITLE
planner: fix index-out-of-range error when checking only_full_group_by

### LINKED ISSUE #23839: TiDB throw error when explain a complex sql
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> show create table bb;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| BB    | CREATE TABLE `BB` (
  `col_int` int(11) DEFAULT NULL,
  `col_varchar_10` varchar(10) DEFAULT NULL,
  `pk` int(11) NOT NULL AUTO_INCREMENT,
  `col_int_not_null` int(11) NOT NULL,
  `col_decimal` decimal(10,0) DEFAULT NULL,
  `col_datetime` datetime DEFAULT NULL,
  `col_decimal_not_null` decimal(10,0) NOT NULL,
  `col_datetime_not_null` datetime NOT NULL,
  `col_varchar_10_not_null` varchar(10) NOT NULL,
  PRIMARY KEY (`pk`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=2000001 |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> explain SELECT OUTR . col2 AS X FROM (SELECT INNR . col1 as col1, SUM( INNR . col2 ) as col2 FROM (SELECT INNR . `col_int_not_null` + 1 as col1, INNR . `pk` as col2 FROM BB AS INNR) AS INNR GROUP BY col1) AS OUTR2 INNER JOIN (SELECT INNR . col1 as col1, MAX( INNR . col2 ) as col2 FROM (SELECT INNR . `col_int_not_null` + 1 as col1, INNR . `pk` as col2 FROM BB AS INNR) AS INNR GROUP BY col1) AS OUTR ON OUTR2.col1 = OUTR.col1 GROUP BY OUTR . col1, OUTR2 . col1 HAVING X <> 'b';
ERROR 1105 (HY000): runtime error: index out of range [-1]
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

the query plan
### 3. What did you see instead (Required)

runtime error
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                            |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-2548-ge2a296144
Edition: Community
Git Commit Hash: e2a296144097273d97f31001cbfbdafde975769e
Git Branch: master
UTC Build Time: 2021-04-01 17:02:46
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #23839 

Problem Summary:
TiDB throws  index-out-of-range error when checking only_full_group_by for a complex sql.

### What is changed and how it works?

What's Changed & How it Works:

In function `checkColFuncDepend`, if `expression.FindFieldName` returns `-1`, i.e., it doesn't find the column name, we just jump out of loop.

However, the problem is not entirely solved. Currently TiDB thinks the sql in #23839 violates only_full_group_by rule but MySQL doesn't think so. As https://github.com/pingcap/tidb/issues/22301#issuecomment-810905725 says, `only-full-group-by checker` will be refactored in the future.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- fix index-out-of-range error when checking only_full_group_by

### CODE REVIEW

**Review by Reminiscent — APPROVED** [review]
/LGTM

**Review by lzmhhh123 — APPROVED** [review]
Should this PR be cherry-picked to `5.0` or `4.0`?

**Comment by zimulala** [comment]
> Should this PR be cherry-picked to `5.0` or `4.0`?

@xuyifangreeneyes

**Comment by xuyifangreeneyes** [comment]
/needs-cherry-pick-5.0

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2676,7 +2676,7 @@ func checkColFuncDepend(
 			Name:   colInfo.Name,
 		}
 		pIdx, err := expression.FindFieldName(p.OutputNames(), pkName)
-		if err != nil {
+		if err != nil || pIdx < 0 {
```
Comment (zhangjinpeng1987) [inline planner/core/logical_plan_builder.go:2679]: Better to add a comment to explain when `pIdx` will be less than 0.

Comment (xuyifangreeneyes) [reply planner/core/logical_plan_builder.go:2679]: Here is a case when `pIdx` will be less than 0.
```
CREATE TABLE `BB` (
  `pk` int(11) NOT NULL AUTO_INCREMENT,
  `col_int_not_null` int NOT NULL,
  PRIMARY KEY (`pk`)
);

SELECT OUTR . col2 AS X
FROM
  BB AS OUTR2
INNER JOIN
  (SELECT col_int_not_null AS col1,
          pk AS col2
   FROM BB) AS OUTR ON OUTR2.col_int_not_null = OUTR.col1
GROUP BY OUTR2.col_int_not_null;
```
When we `checkOnlyFullGroupBy` for the SELECT statement and enter `checkColFuncDepend`, `pkName.Table` is `OUTR` which is an alias, while `pkName.Name` is `pk` which is a original name. Hence `expression.FindFieldName` will fail and `pIdx` will be less than 0. Currently, when we meet `pIdx < 0`, we directly regard `primaryFuncDepend` as `false` and jump out. This way is easy to implement but makes `only-full-group-by checker` not smart enough. Later we will refactor `only-full-group-by checker` and resolve the inconsistency between the alias table name and the original column name.

**Pull request merged**